transports/janus_mqtt.c

```diff
@@ -214,17 +235,34 @@ void janus_mqtt_client_subscribe_success_impl(void *context);
 void janus_mqtt_client_subscribe_failure_impl(void *context, int rc);
 void janus_mqtt_client_admin_subscribe_success_impl(void *context);
 void janus_mqtt_client_admin_subscribe_failure_impl(void *context, int rc);
-void janus_mqtt_client_publish_janus_success_impl(void *context);
-void janus_mqtt_client_publish_janus_failure_impl(void *context, int rc);
-void janus_mqtt_client_publish_admin_success_impl(void *context);
-void janus_mqtt_client_publish_admin_failure_impl(void *context, int rc);
-void janus_mqtt_client_publish_status_success_impl(void *context);
-void janus_mqtt_client_publish_status_failure_impl(void *context, int rc);
+void janus_mqtt_client_publish_janus_success_impl(char *topic);
+void janus_mqtt_client_publish_janus_failure_impl(int rc);
+void janus_mqtt_client_publish_admin_success_impl(char *topic);
+void janus_mqtt_client_publish_admin_failure_impl(int rc);
+void janus_mqtt_client_publish_status_success_impl(char *topic);
+void janus_mqtt_client_publish_status_failure_impl(int rc);
 
 /* We only handle a single client */
 static janus_mqtt_context *context_ = NULL;
 static janus_transport_session *mqtt_session = NULL;
 
+#ifdef MQTTVERSION_5
+/* MQTT 5 specific statics and functions */
+static GHashTable *janus_mqtt_transaction_states;
+static GRWLock janus_mqtt_transaction_states_lock;
```

Comment:
Why this kind of lock instead of the regular `janus_mutex`?

Reply:
Because it allows concurrent reading which seems to be more effective than mutex in this case.